File: mbt/lib/rust/src/runner.rs
use crate::plugin_service::FizzBeeServiceImpl;
use crate::pb::fizz_bee_mbt_plugin_service_server::FizzBeeMbtPluginServiceServer;
use crate::error::MbtError;
use crate::traits::{Model, DispatchModel};
use crate::config::TestOptions;
use tokio::runtime::Builder;
use tonic::transport::Server;

// Imports for UDS and Child Process
use tokio::net::UnixListener;
use tokio_stream::wrappers::UnixListenerStream;
use std::path::Path;
use std::fs;
use tokio::process::Command;
use std::env;
use tokio::signal::unix::{signal, SignalKind};

// Helper for Binary Path
fn get_mbt_bin_path() -> String {
    env::var("FIZZBEE_MBT_BIN").unwrap_or_else(|_| "fizzbee-mbt-runner".to_string())
}

/// Checks environment variable first, then TestOptions, otherwise returns None.
fn get_arg_value(env_var: &str, option_value: Option<u32>) -> Option<String> {
    // Precedence 1: Environment variable
    if let Ok(val) = env::var(env_var) {
        Some(val)
    }
    // Precedence 2: TestOptions
    else if let Some(val) = option_value {
        Some(val.to_string())
    }
    // Fallback: None
    else {
        None
    }
}

// UDS Socket Path Constant
const UDS_PATH: &str = "/tmp/fizzbee_mbt.sock";

// --- Private Helper to start the Server Future ---
/// Binds to the UDS socket and returns the Server future.
fn serve_uds_socket<D>(
    dispatcher: D,
) -> Result<impl std::future::Future<Output = Result<(), tonic::transport::Error>>, MbtError>
where
    D: Model + DispatchModel + Send + Sync + 'static,
{
    let path = Path::new(UDS_PATH);

    // Clean up the old socket file if it exists.
    if path.exists() {
        fs::remove_file(path).map_err(MbtError::from_err)?;
    }

    let uds = UnixListener::bind(path).map_err(MbtError::from_err)?;
    let incoming = UnixListenerStream::new(uds);

    let service = FizzBeeServiceImpl::new(dispatcher);

    Ok(
        Server::builder()
            .add_service(FizzBeeMbtPluginServiceServer::new(service))
            .serve_with_incoming(incoming)
    )
}

// --- 3. Public Entry Point for Running Tests (MBT Orchestrator) ---
/// Starts the gRPC server, launches the child process, and waits for both to complete.
/// Returns the exit status of the child process.
pub fn run_mbt_test<D>(
    dispatcher: D,
    options: TestOptions,
) -> Result<(), MbtError>
where
    D: Model + DispatchModel + Send + Sync + 'static,
{
    let rt = Builder::new_multi_thread()
        .enable_all()
        .build()
        .map_err(MbtError::from_err)?;

    rt.block_on(async {
        // 1. Get the server future
        let server_future = serve_uds_socket(dispatcher)?;

        // 2. Launch the child process after the server has successfully bound.
        let bin_path = get_mbt_bin_path();
        let mut args: Vec<String> = Vec::new();

        // --- 2a. Mandatory Plugin Address ---
        // NOTE: Uses 'unix://' prefix, common for MBT runners to identify UDS
        args.push(format!("--plugin-addr={}", UDS_PATH));

        // --- 2b. Optional Config Options (Env > Options) ---
        if let Some(val) = get_arg_value("MAX_ACTIONS", options.max_actions) {
            args.push(format!("--max-actions={}", val));
        }

        if let Some(val) = get_arg_value("MAX_SEQ_RUNS", options.max_seq_runs) {
            args.push(format!("--max-seq-runs={}", val));
        }

        if let Some(val) = get_arg_value("MAX_PARALLEL_RUNS", options.max_parallel_runs) {
            args.push(format!("--max-parallel-runs={}", val));
        }

        // --- 2c. Optional Seed Options (Env only) ---
        if let Ok(val) = env::var("SEQ_SEED") {
            args.push(format!("--seq-seed={}", val));
        }

        if let Ok(val) = env::var("PARALLEL_SEED") {
            args.push(format!("--parallel-seed={}", val));
        }

        let mut child = Command::new(&bin_path)
            .args(&args)
            .spawn()
            .map_err(MbtError::from_err)?;

        let child_future = child.wait();

        // 3. Setup signal handlers for graceful exit (SIGINT/Ctrl+C and SIGTERM)
        let mut sigint = signal(SignalKind::interrupt()).map_err(MbtError::from_err)?;
        let mut sigterm = signal(SignalKind::terminate()).map_err(MbtError::from_err)?;

        // 4. Concurrently wait for the server, child process, or an interrupt signal.
        tokio::select! {
            // Wait for the gRPC server to terminate
            res = server_future => {
                let _ = child.kill().await;
                match res {
                    Ok(_) => {
                        Err(MbtError::other("gRPC server terminated unexpectedly before child process."))
                    },
                    Err(e) => {
                        Err(MbtError::other(format!("gRPC server failed: {}", e)))
                    }
                }
            }
            // Wait for the child process to terminate
            res = child_future => {
                match res {
                    Ok(status) => {
                        if status.success() {
                            Ok(())
                        } else {
                            Err(MbtError::other(format!("Child process failed with status: {}", status)))
                        }
                    }
                    Err(e) => {
                        Err(MbtError::from_err(e))
                    }
                }
            }
            // Handle interruption signals
            _ = sigint.recv() => {
                // Kill the child process
                let _ = child.kill().await;
                child.wait().await.map_err(MbtError::from_err)?; // Wait for cleanup
                Err(MbtError::other(format!("Test interrupted by signal {:?}", SignalKind::interrupt())))
            }
            _ = sigterm.recv() => {
                // Kill the child process
                let _ = child.kill().await;
                child.wait().await.map_err(MbtError::from_err)?; // Wait for cleanup
                Err(MbtError::other(format!("Test interrupted by signal {:?}", SignalKind::terminate())))
            }
        }
    })
}

File: mbt/lib/typescript/src/runner.ts
import * as grpc from '@grpc/grpc-js';
import { spawn, ChildProcess } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { FizzBeeMbtPluginService } from './plugin-service';
import type { Model, ActionFunc } from './interfaces';
import { FizzBeeMbtPluginServiceService } from '../proto-gen/mbt_plugin_grpc_pb';

/**
 * Options for configuring the test runner.
 */
export interface RunTestsOptions {
  'max-seq-runs'?: number;
  'max-parallel-runs'?: number;
  'max-actions'?: number;
  'seq-seed'?: number;
  'parallel-seed'?: number;
  [key: string]: any;
}

/**
 * Gets a numeric value from environment variable.
 */
function getEnvNumber(key: string): number | undefined {
  const value = process.env[key];
  if (value) {
    const num = parseInt(value, 10);
    return isNaN(num) ? undefined : num;
  }
  return undefined;
}

/**
 * Gets the path to the fizzbee-mbt-runner binary from environment.
 */
function getMbtBinPath(): string {
  return process.env.FIZZBEE_MBT_BIN || 'fizzbee-mbt-runner';
}

/**
 * Main function to run MBT tests.
 */
export async function runTests(
  model: Model,
  actionsRegistry: Map<string, Map<string, ActionFunc>>,
  options: RunTestsOptions = {}
): Promise<void> {
  // Create temporary directory for socket
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'fizzbee-mbt-'));
  const socketPath = path.join(tmpDir, 'plugin.sock');

  let server: grpc.Server | null = null;
  let childProcess: ChildProcess | null = null;

  try {
    // Create service implementation
    const serviceImpl = new FizzBeeMbtPluginService(model, actionsRegistry);

    // Create and start gRPC server
    server = new grpc.Server();
    server.addService(FizzBeeMbtPluginServiceService, serviceImpl);

    // Start server on Unix domain socket
    await new Promise<void>((resolve, reject) => {
      server!.bindAsync(
        `unix://${socketPath}`,
        grpc.ServerCredentials.createInsecure(),
        (error) => {
          if (error) {
            reject(error);
          } else {
            console.log(`gRPC server started on ${socketPath}`);
            resolve();
          }
        }
      );
    });

    // Build command arguments
    const runnerCmd = getMbtBinPath();
    const args = [`--plugin-addr=${socketPath}`];

    // Add options, with environment variable overrides
    for (const [optionName, optionValue] of Object.entries(options)) {
      let value = optionValue;

      // Override with environment variables if present
      if (optionName === 'max-actions') {
        value = getEnvNumber('FIZZBEE_MBT_MAX_ACTIONS') ?? value;
      } else if (optionName === 'max-seq-runs') {
        value = getEnvNumber('FIZZBEE_MBT_MAX_SEQ_RUNS') ?? value;
      } else if (optionName === 'max-parallel-runs') {
        value = getEnvNumber('FIZZBEE_MBT_MAX_PARALLEL_RUNS') ?? value;
      }

      args.push(`--${optionName}=${value}`);
    }

    // Add seed flags from environment if provided
    const seqSeed = getEnvNumber('FIZZBEE_MBT_SEQ_SEED');
    if (seqSeed !== undefined) {
      args.push(`--seq-seed=${seqSeed}`);
    }
    const parallelSeed = getEnvNumber('FIZZBEE_MBT_PARALLEL_SEED');
    if (parallelSeed !== undefined) {
      args.push(`--parallel-seed=${parallelSeed}`);
    }

    // Start child process
    console.log(`Starting runner: ${runnerCmd} ${args.join(' ')}`);
    childProcess = spawn(runnerCmd, args, {
      stdio: 'inherit'
    });

    // Handle process signals
    const signalHandler = () => {
      console.log('Interrupt received, stopping runner and test plugin...');
      if (childProcess) {
        childProcess.kill('SIGTERM');
      }
    };

    process.on('SIGINT', signalHandler);
    process.on('SIGTERM', signalHandler);

    // Wait for child process to complete
    await new Promise<void>((resolve, reject) => {
      childProcess!.on('exit', (code, signal) => {
        if (code === 0) {
          console.log('Runner exited successfully');
          resolve();
        } else {
          reject(new Error(`Runner exited with code ${code}, signal ${signal}`));
        }
      });

      childProcess!.on('error', (error) => {
        reject(new Error(`Failed to start runner: ${error.message}`));
      });
    });

  } finally {
    // Graceful shutdown
    if (server) {
      const serverRef = server; // Capture for use in closures
      await new Promise<void>((resolve) => {
        const shutdownTimeout = setTimeout(() => {
          console.log('Forcing server stop');
          serverRef.forceShutdown();
          resolve();
        }, 5000);

        serverRef.tryShutdown(() => {
          clearTimeout(shutdownTimeout);
          console.log('Test executor shut down gracefully');
          resolve();
        });
      });
    }

    // Clean up temporary directory
    try {
      if (fs.existsSync(socketPath)) {
        fs.unlinkSync(socketPath);
      }
      fs.rmdirSync(tmpDir);
    } catch (error) {
      // Ignore cleanup errors
    }
  }
}
